sklearn/utils/testing.py

```diff
@@ -372,14 +375,52 @@ def __exit__(self, *exc_info):
     assert_greater = _assert_greater
 
 
+def _sparse_dense_allclose(val1, val2, rtol=1e-7, atol=0):
```

Comment:
I should say: one reason I prefer an assert is that an error message should be passed out if this function.